text/0000-partial_types.md

```diff
@@ -0,0 +1,395 @@
+```rust
+impl Point {
+    fn ref_x (self : & Self.{x}) -> &f64 {
+        &self.x
+    }
+
+    fn refmut_y (self : &mut Self.{y}) -> &mut f64 {
+        &mut self.y
+    }
+}
+let ref_p1x = p1.ref_x();
+let refmut_p1y = p1.refmut_y();
+```
+It is expected, that `self` is **always** cut partiality of argument by same partiality as self-parameter by partial expression before use (even if implicit rules are off)!
+
+Pseudo-rust:
+```rust
+fn ref_xy (self : & Self.{ x, y}) -> &f64 {
+    /*  */
+}
+
+p1.ref_xy();
+// which "desugar"
+Point::ref_xy(& p1.{x, y});
+```
+
+Product-Typed argument type must match with function parameter type or argument type could has **more** permitted partiality then parameter type.
+```rust
+// Struct ~ Product Type
+struct S4 {a : i32, b : i32, c : i32, d : i32}
+
+fn do_sab(s : S4.{a, b}) { /* .. */ }
+
+let s = S4 {a: 6, b: 7, c: 8, d: 9};
+
+do_sab(s.{a});       // s.{a} - error
+do_sab(s.{b});       // s.{b} - error
+do_sab(s.{a, b});    // s.{a, b} - Ok
+do_sab(s.{a, b, c}); // s.{a, b, c} - Ok
+do_sab(s);           // s.{*} - Ok
+```
+
+
+# Reference-level explanation
+
+The core Idea of this proposal is "Proxy Borrowing" - we borrow the whole variable, but borrow-checker pretends it borrow just permitted/allowed fields.
+
+Automatically Type-checker gives a mathematical guarantee, because all denied/forbidden fields remain intact! 
+
+And this mean, that Proxy Borrowing borrowing is fully **safe** and _zero cost_ in binary.
+
+## Proxy Borrowing
+
+Borrowing rules for partial types:
+
+`PermittedField` field borrowing rules are ordinary Rust rules. New variable borrows the whole variable (with partial type), but checker pretends it borrows just permitted fields of this variable.
+
+Not-`PermittedField` filed is always is ready to borrow regardless if origin field is denied(by move, by reference, by borrow).
+
+When we write a code for full or partial borrow, the link of object itself returns, but borrow-checker checks to borrow of permitted fields only.
+
+This new mechanism of is simple and universal.
+
+```rust
+struct S4 {a : i32, b : i32, c : i32, d : i32}
+let s = S4 {a : 5, b: 6, c: 7, d: 8};
+    // s : S4
+
+let r_sd = & s.{d};
+    // r_sd : & S4.{d}
+    //
+    // borrow-checker check just for &s.d
+
+let mut mr_sabc = &mut s.{a, b, c};
+    // mr_sabc : &mut S4.{a, b, c}
+    //
+    // borrow-checkercheck just for &mut s.a, &mut s.b, &mut s.c
+
+let rr_sbc = & mr_sabc.{b, c};
+    // rr_sbc : && S4.{b, c}
+    //
+    // borrow-checker check just for &mr_sabc.b, &mr_sabc.c
+
+let mut mrr_sa = &mut mr_sabc.{a};
+    // mrr_sa : &&mut S4.{a}
+    //
+    // borrow-checker check just for &mut mr_sabc.a
+```
+
+## Syntax
+
+Second, but still important - syntax.
+
+### Partiality Syntax
+
+Minimal Partiality we could write:
+```
+Partiality:      .{ PartialFields* }
+PartialFields:   PermittedField (, PermittedField )* ,?
+PermittedField:  IDENTIFIER | TUPLE_INDEX
```

Comment:
what of you had:
```rust
struct Foo {
    a: i32,
    bar: Bar,
}

struct Bar {
    b: f32,
    c: String,
}
```

and you wanted to borrow both `a` and `bar.b` but not `bar.c`?

I think extending partial references to allow this would be useful:
```rust
impl Foo {
    fn baz(&self.{a, bar.c}) {
    }
}
```

Reply:
*I agree. I'll update this proposal to allow sub-partiality:*

Minimal Partiality we could write:
```
Partiality:      .{ PartialFields* }
PartialFields:   PartialField (, PartialField )* ,?
PartialField:    PermittedField
PermittedField:  IDENTIFIER | TUPLE_INDEX
```

If we wish to describe partial structs with partial structs inside, we must have a bit more complex Partiality:

```
PartialField:    PermittedField Partiality?
```